language/tools/test_generation/src/abstract_state.rs

```diff
@@ -2,36 +2,77 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use std::collections::HashMap;
-use vm::file_format::SignatureToken;
+use vm::file_format::{Kind, SignatureToken};
 
 /// The BorrowState denotes whether a local is `Available` or
 /// has been moved and is `Unavailable`.
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Debug, Clone, Copy, PartialEq)]
 pub enum BorrowState {
     Available,
     Unavailable,
 }
 
+/// This models a value on the stack or in the locals
+#[derive(Debug, Clone, PartialEq)]
+pub struct AbstractValue {
+    /// Represents the type of the value
+    pub token: SignatureToken,
+
+    /// Represents the kind of the value
+    pub kind: Kind,
+}
+
+impl AbstractValue {
+    /// Create a new primitive `AbstractValue` given its type; the kind will be `Unrestricted`
+    pub fn new_primitive(token: SignatureToken) -> AbstractValue {
+        assert!(
```

Comment:
This looks like a precondition.

Reply:
This has been changed to a precondition, along with other checks and asserts, in an upcoming PR that includes changes from MIRAI and general safety improvements for the tool.